pghoard/rohmu/object_storage/s3.py

```diff
@@ -62,152 +77,245 @@ def get_metadata_for_key(self, key):
+            )
+            for item in response.get("Contents", []):
+                if with_metadata:
+                    metadata = self._metadata_for_key(item["Key"])
+                else:
+                    metadata = None
+                name = self.format_key_from_backend(item["Key"])
+                yield {
+                    "last_modified": item["LastModified"],
+                    "metadata": metadata,
+                    "name": name,
+                    "size": item["Size"],
+                }
+            if "NextContinuationToken" in response:
+                continuation_token = response["NextContinuationToken"]
             else:
-                metadata = None
-            yield {
-                "last_modified": parse_timestamp(item.last_modified),
-                "metadata": metadata,
-                "name": name,
-                "size": item.size,
-            }
+                break
 
-    def get_contents_to_file(self, key, filepath_to_store_to, *, progress_callback=None):
+    def _get_object_stream(self, key):
         key = self.format_key_for_backend(key)
-        item = self.bucket.get_key(key)
-        if item is None:
-            raise FileNotFoundFromStorageError(key)
-        item.get_contents_to_filename(filepath_to_store_to, cb=progress_callback, num_cb=1000)
-        return item.metadata
+        try:
+            response = self.s3_client.get_object(
+                Bucket=self.bucket_name,
+                Key=key,
+            )
+        except botocore.exceptions.ClientError as ex:
+            status_code = ex.response.get("ResponseMetadata", {}).get("HTTPStatusCode")
+            if status_code == 404:
+                raise FileNotFoundFromStorageError(key)
+            else:
+                raise StorageError("Fetching the remote object {} failed".format(key)) from ex
+        return response["Body"], response["ContentLength"]
+
+    def _read_object_to_fileobj(self, fileobj, streaming_body, body_length, cb=None):
+        data_read = 0
+        reported_pct = 0
+        while data_read < body_length:
+            read_amount = body_length - data_read
+            if read_amount > READ_BLOCK_SIZE:
+                read_amount = READ_BLOCK_SIZE
+            data = streaming_body.read(amt=read_amount)
+            fileobj.write(data)
+            data_read += len(data)
+            if cb:
+                download_pct = data_read * 100 // body_length
+                if download_pct > reported_pct:
+                    cb(download_pct, 100)
+                    reported_pct = download_pct
+        if cb:
+            cb(100, 100)
+
+    def get_contents_to_file(self, key, filepath_to_store_to, *, progress_callback=None):
+        with open(filepath_to_store_to, "wb") as fh:
+            stream, length = self._get_object_stream(key)
+            self._read_object_to_fileobj(fh, stream, length, cb=progress_callback)
+        return self.get_metadata_for_key(key)
```

Comment:
do we really have to issue a second HTTP request to get the metadata here?  This will slow down WAL recovery when we need to do two http requests for every segment